### ISSUE TITLE
Returned routes not 

### ISSUE THREAD

**Issue by szymone** [valhalla/valhalla#3680]
Hi!
I've come across a strange issue with routes nearby gates with no access. The returned shapes don't overlap with osm ways / graph edges. This is related to minimum_reachability value and setting it to 0 fix this, but causes route not found errors in other cases. 

![image](https://user-images.githubusercontent.com/10035692/178581421-9c7dd729-1cb0-4acf-852c-d9aee130195c.png)
There is no osm way under line pointed with arrow.

Some examples:

https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=22.074124498612946%2C48.963925693746056%2C22.07842841271486%2C48.96395579152026%2C22.0781521%2C48.9633729

https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=21.9011069%2C48.7637972%2C21.901487245813772%2C48.76404976263312%2C21.9012539%2C48.7645005

https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=18.060375864915844%2C50.11149129237608%2C18.0604636%2C50.1113051%2C18.0602043%2C50.1113078

https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=21.1943144%2C49.5143381%2C21.1946386%2C49.5143172%2C21.1945903%2C49.5152148

Any ideas what can cause this and how to resolve this easily?

**Comment by szymone** [comment]
@kevinkreiser, @dnesbitt61, @nilsnolde Any comments or ideas on this? I've found another example where route is found and jumps from the cliff.

https://valhalla.openstreetmap.de/directions?profile=pedestrian&wps=14.59671007542056%2C53.9844032592876%2C14.589159383322114%2C53.98627151150229%2C14.581711204518799%2C53.984253632160225

<img width="547" alt="image" src="https://user-images.githubusercontent.com/10035692/179964679-094a454d-8357-4927-bb70-b0c0f8689f0d.png">

**Comment by kevinkreiser** [comment]
Looks like a pretty serious but to me but I'd hate to guess what it could be without looking. We just need time to debug it. Sadly I have 4 other valhall issues I'm tracking at the moment so this one will have to join the queue despite its serious nature.

**Comment by dnesbitt61** [comment]
This seems to be a bug in the web app. The middle location has no type set, so it defaults to "break". Take the URL (using web tools) and send directly to the server - you get 2 legs in the response. I suspect the web app here is connecting the geometry of 2 legs together (hence the straight line from the end of the path to location 2 in the labeled diagram above). If the middle location is set to "type":"via" then only 1 leg is returned.

https://valhalla1.openstreetmap.de/route?json={"costing":"pedestrian","costing_options":{"pedestrian":{"exclude_polygons":[],"use_ferry":1,"service_penalty":15,"service_factor":1,"shortest":false,"use_hills":0.5,"walking_speed":5.1,"walkway_factor":1,"sidewalk_factor":1,"alley_factor":2,"driveway_factor":5,"step_penalty":0,"max_hiking_difficulty":1}},"exclude_polygons":[],"locations":[{"lon":14.595776796340944,"lat":53.98471102429982},{"lon":14.589159383322114,"lat":53.98627151150229},{"lon":14.581711204518799,"lat":53.984253632160225}],"directions_options":{"units":"kilometers"},"id":"valhalla_directions"}

**Comment by dnesbitt61** [comment]
I suspect the initial cases in this issue are the same. Perhaps an easy fix is to make sure only the first and last locations are type=break, and any intermediate locations are set to "via", "through" in the web app. There is no Valhalla bug here.

**Comment by nilsnolde** [comment]
Ayyye damn, sorry about that.. I’m on a week vacation from tonight, if @TimMcCauley didn't fix it until then, I will. Or someone can put a PR to https://github.com/gis-ops/valhalla-app

**Comment by szymone** [comment]
Thanks for your comments! @dnesbitt61 This is not related to web app and setting location type to via doesn't resolve this issue. In that case one route leg is returned, but the shape is the same and invalid.

It seems that this error is related to connecting routes found between every two locations in thor. When minimum_reachability > 0 or radius > 0, locations can have multiple correlation edges used later for routing and the resulting route contains two route segments found between different correlations. See picture below. 

![image](https://user-images.githubusercontent.com/10035692/179989667-306962b1-0718-4748-90be-b7d4bf8ad292.png)

If you set minimum_reachability and radius to 0, every location has only one correlation edge so the resulting route segments are connected correctly.

**Comment by dnesbitt61** [comment]
I must have had old data / code. I pulled latest updates and rebuilt Valhalla and data for Poland and I now see the issue. I logged the shape as polyline5 encoded and draw it in geojson.io and it looks a bit different but shows the issue. 

I would say that the web app still has a bug - if each location is type=break then each leg's geometry / shape should be drawn separately (not connected). I would use type=via or type=through on intermediate locations though.

**Comment by kevinkreiser** [comment]
yeah this is a long standing issue with multipoint routes. if we have more than one candidate edge for a given location it is possible that the second leg of the route might use a different candidate than the first one to fulfill the reachability criteria. the reason `through` works is because it forces you to use the same edge, but because breaks and vias can make uturns at nodes its possible to use different but connected edges. it seems to me that we could fail the route if the edges arent connected by simply culling the candidates of the location as we try to do the second leg of the route. i swore i did this already but maybe there is an edge case that missed.
